LuaD enhancements

TurkeyMan opened the pull request:
Fixed static arrays.
Fixed support for calling functions which receive const args.
Rework struct to use userdata; conversion is not lossy, conversion should be faster, less garbage, supports more features (methods, properties).
Classes improved to support fields, properties.
Static type interfaces and creation of struct instances in Lua code.
Default construction is supported in static type interfaces in lieu of a constructor.

TurkeyMan committed:
struct changed to use userdata blocks. Structs are handled by-ref in Lua code to match handling of regular tables.

struct and class have uniform support for: methods, properties, member variable accessors.
Refactor common code into helpers.d
Added support for const function args.
pushMethod enhanced to support struct's aswell.
New alias syntax updates.

TurkeyMan committed:
Added support for enum's.

Enums are held in Lua as strings.
No support for enum values that are not valid keys.
No support for bitfields (yet?).
Conversion function performance could be improved (linear search >_<).

TurkeyMan committed:
local cruft

TurkeyMan committed:
Added @noscript

JakobOvrum commented on luad/conversions/arrays.d:
```diff
@@ -27,7 +27,16 @@ T getArray(T)(lua_State* L, int idx) if (isArray!T)
+		if(len != T.length)
+			luaL_error(L, "Incorrect number of array elements: %d, expected: %d", len, T.length);
+
+		T arr;
```
Maybe we should void-initialize this? To preserve `@safe`ty, the void-initialized path can be a specialization for element types without indirection (std.traits.hasIndirections should do the trick).

JakobOvrum commented on luad/conversions/classes.d:
```diff
@@ -165,7 +244,12 @@ void pushStaticTypeInterface(T)(lua_State* L)
 		static if(isStaticMember!(T, member))
 		{
 			enum isFunction = is(typeof(mixin("T." ~ member)) == function);
+			static if(isFunction)
```
Static if not needed; it's all handled in the then-clause

TurkeyMan replied:
Actually, it is, I just didn't remove it from the 'then' expression.
The following expression (function args) can't compile if !isFunction, so it needs static if.

JakobOvrum commented on luad/conversions/enums.d:
```diff
@@ -0,0 +1,190 @@
+
+immutable(KeyValuePair!Enum)[] getKeyValuePairs(Enum)() pure nothrow @nogc
+{
+	static __gshared immutable(KeyValuePair!Enum[]) kvp = [ EnumKeyValuePair!Enum ];
```
__gshared is redundant for immutable globals

TurkeyMan replied:
Is it? I've never been able to find solid information about the complete set of combinations... I just over-specify attributes to make sure :)

JakobOvrum commented on luad/conversions/enums.d:
```diff
@@ -0,0 +1,190 @@
+//		lua_pushnil(L);
+
+		// error?
+		luaL_error(L, "invalid value for enum type %s", T.stringof.ptr);
```
Erroring seems like the most forward-compatible thing to do for now. We can loosen the restriction later if evidence suggests it makes sense to do so

TurkeyMan replied:
I agree... hence the code :)
I'll remove the other junk, I was really just presenting the option to you.

JakobOvrum commented on luad/conversions/enums.d:
```diff
@@ -0,0 +1,190 @@
+{
+	// TODO: check to see if idx is a number, if it is, convert it directly?
+
+	size_t l;
```
Please don't use single-letter variables named `l` as depending on the font, it can look extremely similar to `1`

TurkeyMan replied:
Haha, righto.

JakobOvrum commented on luad/conversions/enums.d:
```diff
@@ -0,0 +1,190 @@
+	return getEnumValue!T(L, s[0..l]);
+}
+
+void pushStaticTypeInterface(T)(lua_State* L) if(is(T == enum))
```
Integral enums also have the properties `min` and `max`

TurkeyMan replied:
Good point.

JakobOvrum commented on luad/conversions/enums.d:
```diff
@@ -0,0 +1,190 @@
+		lua_rawseti(L, -2, i+1);
+	}
+
+	lua_setfield(L, -3, "values");
```
These aren't D properties, so it's possible that the enum has members named `keys` or `values`, which would collide with this :(

TurkeyMan replied:
Yeah, I already considered this... what do you reckon? I think it's useful to be able to get at these from Lua, since there's not really any other way.

JakobOvrum commented on luad/conversions/functions.d:
```diff
@@ -162,20 +170,23 @@ extern(C) int methodWrapper(T, Class, bool virtual)(lua_State* L)
 	else
 	{
-		Args allArgs;
+		UnqualTuple!(Args) allArgs;
```
`UnqualTuple!Args`

TurkeyMan replied:
This has all changed drastically in my uncommitted code...

JakobOvrum commented on luad/conversions/functions.d:
```diff
@@ -249,18 +260,26 @@ void pushFunction(T)(lua_State* L, T func) if (isSomeFunction!T)
+void pushMethod(T, string member)(lua_State* L) if (isSomeFunction!(__traits(getMember, T, member)))
 {
-	alias typeof(mixin("&Class.init." ~ member)) T;
+	alias typeof(mixin("&T.init." ~ member)) M;
```
Let's opportunistically use the new alias syntax whenever possible

TurkeyMan replied:
I was just sticking with the established convention already present throughout the codebase... do you want me to change everywhere I encounter?

JakobOvrum replied:
Yeah, that's what we do in Phobos. Maybe soon we can use `dfix` to do it all in one swoop (although LuaD's codebase is relatively small, so maybe 10 minutes with grep is all that's needed...)

JakobOvrum commented on luad/conversions/functions.d:
```diff
@@ -249,18 +260,26 @@ void pushFunction(T)(lua_State* L, T func) if (isSomeFunction!T)
-	alias typeof(mixin("&Class.init." ~ member)) T;
+	alias typeof(mixin("&T.init." ~ member)) M;
+
+	enum isVirtual = is(T == struct) ? false : true;
```
`enum isVirtual = !is(T == struct);`

edit:

Of course, it should also check for `final` but that was a problem with the original code as well.

JakobOvrum commented on luad/conversions/helpers.d:
```diff
@@ -0,0 +1,179 @@
+	enum isInternal = field.length >= 2 && field[0..2] == "__";
+}
+
+template isOperator(string field)
```
This does not support the old overload names that the compiler still supports, like `opAdd` etc. Not sure if we wanna bother supporting those though.

TurkeyMan replied:
I'd say no. I'm sure there are a couple of others that I missed too.
This whole block of code seems really horrible to me... it feels pretty lame excluding a whole bunch of magic function names. I couldn't think of a better way though.

JakobOvrum commented on luad/conversions/helpers.d:
```diff
@@ -0,0 +1,179 @@
+	                  field == "opDispatch";
+}
+
+template skipField(T, string field)
```
`skipMember`?

JakobOvrum commented on luad/conversions/helpers.d:
```diff
@@ -0,0 +1,179 @@
+template isProperty(T, string member)
+{
+	static if(isMemberFunction!(T, member))
+		enum isProperty = functionAttributes!(mixin("T.init." ~ member)) & FunctionAttribute.property;
```
`&&` short-circuits, so this could use `enum isProperty = isMemberFunction!(T, member) && ...`?

TurkeyMan replied:
I tried, it doesn't work. I found myself having to do this in lots of places because && didn't short-circuit as expected.

JakobOvrum commented on luad/conversions/helpers.d:
```diff
@@ -0,0 +1,179 @@
+{
+	static if(is(T == class))
+	{
+		class S
```
Why are these empty wrapper types needed?

TurkeyMan replied:
members, properties.
This is quite heavily refactored in my uncommitted code though.

JakobOvrum commented on luad/conversions/structs.d:
```diff
@@ -1,71 +1,170 @@
-Pushing a struct to Lua will create a table and fill it with key-value pairs - corresponding to struct fields - from the struct; the field name becomes the table key as a string.
-Struct methods are treated as if they were delegate fields pointing to the method.
+A struct is treated as a userdata block.
+Members and properties are handled via a thin shim implemented in __index/__newindex.
```
Fields and properties. Methods are members too (hence their other name, member functions).

TurkeyMan replied:
Yeah, I think this is an advanced symptom of some initial cut & pasting ;)

JakobOvrum commented on luad/conversions/structs.d:
```diff
@@ -1,71 +1,170 @@
+import std.traits;
+import std.conv;
+
+private extern(C) int structCleaner(lua_State* L)
```
Isn't this the same as `classCleaner`?

TurkeyMan replied:
It's been refactored to userdataCleaner in helpers.d in my WIP code.

JakobOvrum commented on luad/conversions/structs.d:
```diff
@@ -1,71 +1,170 @@
 
-//TODO: ignore static fields, post-blits, destructors, etc?
-void pushStruct(T)(lua_State* L, ref T value) if (is(T == struct))
+private extern(C) int indexStruct(lua_State* L)
```
Is this different from `indexClass`?

TurkeyMan replied:
It's in helpers.d now.

JakobOvrum commented on luad/conversions/structs.d:
```diff
@@ -1,71 +1,170 @@
+
+ref T getStruct(T)(lua_State* L, int idx) if(is(T == struct))
+{
+	T* udata = cast(T*)lua_touserdata(L, idx);
```
The local variable `udata` is only used once so may as well get rid of it

TurkeyMan replied:
It's so I can breakpoint, step and inspect. I'll tidy everything up before any final merge.

TurkeyMan committed:
Added support for pointers.
Pointers are opaque.
Pointers support 'nil' assignment.
'deref' property on pointers to access the actual object.

TurkeyMan committed:
popValue!(const(T)) can now receive values of T, const(T) and immutable(T).
Added preliminary for class inheritance.

aubade commented on luad/conversions/classes.d:
```diff
@@ -14,97 +15,129 @@ import luad.base;
+		pushMeta!(BaseClassesTuple!T[0])(L);
+		lua_setmetatable(L, -2);
+	}
+
```
^ Is this Particular pushMeta block in any of your feature-specific branches, @TurkeyMan ? My modification to get interfaces to work touches this and I haven't been able to find it elsewhere.

TurkeyMan committed:
Fixed for DMD2.066.1

TurkeyMan committed:
Minor fixes.

GreatEmerald commented:
You forgot to also update the Makefile with the new files you added.

GreatEmerald commented:
Also, somehow this seems to break reading more complex values from Lua scripts, I keep getting the error `userdata expected, got table` whenever trying to either read an array of structs or a struct with complex types inside (functions, other structs that include arrays, etc.)
